test/functional/lib/tests.cpp

```diff
@@ -5308,6 +5320,87 @@ GenericTxMtu()
     TEST_EQUAL(1, Stats.TxInvalidDescriptors);
 }
 
+static
+VOID
+RxIndicate(
+    _In_reads_bytes_(PayloadLength) UCHAR *Payload,
+    _In_ SIZE_T PayloadLength,
+    _In_ MY_SOCKET *Xsk,
+    _In_ NET_IFINDEX IfIndex,
+    _In_ UINT32 QueueId
+    )
+{
+    auto GenericMp = MpOpenGeneric(IfIndex);
+
+    DATA_BUFFER Buffer = {0};
+    Buffer.DataOffset = 0;
+    Buffer.DataLength = (UINT32)PayloadLength;
+    Buffer.BufferLength = Buffer.DataLength;
+    Buffer.VirtualAddress = Payload;
+
+    RX_FRAME Frame;
+    RxInitializeFrame(&Frame, QueueId, &Buffer);
+    TEST_HRESULT(MpRxEnqueueFrame(GenericMp, &Frame));
+    SocketProduceRxFill(Xsk, 1);
+    TEST_HRESULT(TryMpRxFlush(GenericMp));
+}
+
+static
+VOID
+TxIndicate(
+    _In_reads_bytes_(PayloadLength) UCHAR *Payload,
+    _In_ SIZE_T PayloadLength,
+    _In_ MY_SOCKET *Xsk
+    )
+{
+    UINT64 TxBuffer = SocketFreePop(Xsk);
+    UCHAR *TxFrame = Xsk->Umem.Buffer.get() + TxBuffer;
+    UINT32 TxFrameLength = (UINT32)PayloadLength;
+    ASSERT(TxFrameLength <= Xsk->Umem.Reg.ChunkSize);
+    RtlCopyMemory(TxFrame, Payload, PayloadLength);
+
+    UINT32 ProducerIndex;
+    TEST_EQUAL(1, XskRingProducerReserve(&Xsk->Rings.Tx, 1, &ProducerIndex));
+
+    XSK_BUFFER_DESCRIPTOR *TxDesc = SocketGetTxDesc(Xsk, ProducerIndex++);
+    TxDesc->Address.AddressAndOffset = TxBuffer;
+    TxDesc->Length = TxFrameLength;
+    XskRingProducerSubmit(&Xsk->Rings.Tx, 1);
+
+    XSK_NOTIFY_RESULT_FLAGS PokeResult;
+    NotifySocket(Xsk->Handle.get(), XSK_NOTIFY_FLAG_POKE_TX, 0, &PokeResult);
+    TEST_EQUAL(0, PokeResult);
+}
+
+typedef struct {
+    UCHAR *Payload;
+    SIZE_T PayloadLength;
+    MY_SOCKET *Xsk;
+    NET_IFINDEX IfIndex;
+    UINT32 QueueId;
+    BOOLEAN Rx;
+    BOOLEAN Tx;
+} DELAY_INDICATE_THREAD_CONTEXT;
+
+static
+CXPLAT_THREAD_CALLBACK(DelayIndicateFn, Context)
+{
+    DELAY_INDICATE_THREAD_CONTEXT *Ctx = (DELAY_INDICATE_THREAD_CONTEXT *)Context;
+    auto GenericMp = MpOpenGeneric(Ctx->IfIndex);
+
+    Sleep(10);
```

Comment:
This works in kernel? I'm a bit surprised, considering MsQuic has the following for kernel:
```c
_IRQL_requires_max_(PASSIVE_LEVEL)
inline
void
CxPlatSleep(
    _In_ uint32_t DurationMs
    )
{
    CXPLAT_DBG_ASSERT(DurationMs != (uint32_t)-1);

    KTIMER SleepTimer;
    LARGE_INTEGER TimerValue;

    KeInitializeTimerEx(&SleepTimer, SynchronizationTimer);
    TimerValue.QuadPart = Int32x32To64(DurationMs, -10000);
    KeSetTimer(&SleepTimer, TimerValue, NULL);

    KeWaitForSingleObject(&SleepTimer, Executive, KernelMode, FALSE, NULL);
}
```